### ISSUE TITLE
GPT-SoVITS-V4 在RTX5090设备上部署Docker

### ISSUE THREAD

**Issue by lqszhsp** [lqszhsp/GPT-Sovits-v4-DOCKER-for-RTX5090#1]
参考Mark先生提供的方式安装：
[技術探索](https://blog.tarswork.com/zh/category/tech)
[本文重點](https://blog.tarswork.com/zh/post/gpt-sovits-deployment-api-using-docker#chapter1)
[準備 GPT-SoVITS 專案](https://blog.tarswork.com/zh/post/gpt-sovits-deployment-api-using-docker#chapter2)
[建立容器](https://blog.tarswork.com/zh/post/gpt-sovits-deployment-api-using-docker#chapter3)
[測試 API](https://blog.tarswork.com/zh/post/gpt-sovits-deployment-api-using-docker#chapter4)

以下是本文提及的一些重要目錄及檔案:

/opt/GPT-SoVITS
GPT-SoVITS 專案目錄，這是本文假設的路徑，您可以依照自己喜好修改。
/opt/GPT-SoVITS/GPT_weights
訓練好的 GPT_weights 存放目錄。
/opt/GPT-SoVITS/SoVITS_weights
訓練好的 SoVITS_weights 存放目錄。
/opt/GPT-SoVITS/reference
存放 API 會使用到的範例音檔。
/opt/GPT-SoVITS/Dockerfile
建立 Docker 映像檔的 Dockerfile。
/opt/GPT-SoVITS/GPT_SoVITS/configs/tts_infer.yaml
TTS 參數設定。
準備 GPT-SoVITS 專案
# STEP 1
下載 GPT-SoVITS 專案

cd /opt
git clone https://github.com/RVC-Boss/GPT-SoVITS.git
# STEP 2
下載預訓練模型，放置到指定路徑，這邊直接依照 github 上的 README 下載即可。

[RVC-Boss/GPT-SoVITS](https://github.com/RVC-Boss/GPT-SoVITS?tab=readme-ov-file#pretrained-models)
# STEP 3
編輯 /opt/GPT-SoVITS/Dockerfile 檔案(先构建好环境镜像，Python 3.9 | PyTorch 2.8.0dev | CUDA 12.8，我使用的名称是pytorch_python39:latest）：
# ──────────────────────────────────────
# Base CUDA image (from reference)
# ──────────────────────────────────────
FROM pytorch_python39:latest

LABEL maintainer="l67q82@gmail.com"
LABEL version="dev-20250916"
LABEL description="Docker image for GPT-SoVITS"

# Install 3rd party apps
ENV DEBIAN_FRONTEND=noninteractive
ENV TZ=Etc/UTC

# Explicitly set Aliyun mirror for Ubuntu (assuming 24.04 noble)
RUN echo "deb http://mirrors.aliyun.com/ubuntu/ noble main restricted universe multiverse" > /etc/apt/sources.list && \
    echo "deb http://mirrors.aliyun.com/ubuntu/ noble-updates main restricted universe multiverse" >> /etc/apt/sources.list && \
    echo "deb http://mirrors.aliyun.com/ubuntu/ noble-security main restricted universe multiverse" >> /etc/apt/sources.list && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Run apt-get update with retry logic
RUN for i in 1 2 3; do \
        apt-get update && break || { echo "Error: apt-get update failed, retrying ($i/3)"; sleep 5; }; \
    done || { echo "Error: apt-get update failed after 3 retries"; exit 1; }

# Install packages with retry logic
RUN for i in 1 2 3; do \
        apt-get install -y --no-install-recommends \
        tzdata ffmpeg libsox-dev parallel aria2 git git-lfs python3-full python3-venv && \
        break || { echo "Error: apt-get install failed, retrying ($i/3)"; sleep 5; }; \
    done || { echo "Error: apt-get install failed after 3 retries"; exit 1; } && \
    git lfs install && \
    rm -rf /var/lib/apt/lists/*

# Create a virtual environment
RUN python3 -m venv /workspace/venv

# Upgrade pip and install requirements with Tsinghua PyPI mirror
WORKDIR /workspace
COPY requirements.txt /workspace/
RUN /workspace/venv/bin/pip install --no-cache-dir -U pip && \
    /workspace/venv/bin/pip install --no-cache-dir -i https://mirrors.tuna.tsinghua.edu.cn/pypi/web/simple/ --trusted-host mirrors.tuna.tsinghua.edu.cn -r requirements.txt numpy==1.26.4 nltk || { echo "Error: pip install failed"; exit 1; } && \
    /workspace/venv/bin/pip list | grep numpy || { echo "Error: numpy not installed"; exit 1; }

# Define a build-time argument for image type
ARG IMAGE_TYPE=full

# Copy the Docker directory and verify
COPY ./Docker /workspace/Docker
RUN ls -l /workspace/Docker || echo "Warning: /workspace/Docker is empty or missing"

# elite 类型的镜像里面不包含额外的模型
RUN if [ "$IMAGE_TYPE" != "elite" ]; then \
        if [ -f /workspace/Docker/download.sh ]; then \
            chmod +x /workspace/Docker/download.sh && \
            /workspace/Docker/download.sh || echo "Error: download.sh failed"; \
        else \
            echo "Warning: download.sh not found, skipping"; \
        fi && \
        if [ -f /workspace/Docker/download.py ]; then \
            /workspace/venv/bin/python /workspace/Docker/download.py || echo "Error: download.py failed"; \
        else \
            echo "Warning: download.py not found, skipping"; \
        fi && \
        /workspace/venv/bin/python -m nltk.downloader -d /root/nltk_data averaged_perceptron_tagger cmudict || echo "Error: NLTK downloader failed"; \
    fi

# Copy the rest of the application
COPY . /workspace

EXPOSE 9871 9872 9873 9874 9880

# Download NLTK resources during build
RUN /workspace/venv/bin/python -m nltk.downloader -d /root/nltk_data averaged_perceptron_tagger_eng || echo "Error: NLTK downloader for averaged_perceptron_tagger_eng failed"

# Unset proxy for runtime
ENV http_proxy=""
ENV https_proxy=""

# Run the application
CMD ["/workspace/venv/bin/python", "api_v2.py", "-a", "0.0.0.0", "-p", "9880", "-c", "GPT_SoVITS/configs/tts_infer.yaml"]

# STEP 4
編輯 /opt/GPT-SoVITS/GPT_SoVITS/configs/tts_infer.yaml 檔案(change custom to v4)：
custom:
  bert_base_path: GPT_SoVITS/pretrained_models/chinese-roberta-wwm-ext-large
  cnhuhbert_base_path: GPT_SoVITS/pretrained_models/chinese-hubert-base
  device: cuda
  is_half: true
  t2s_weights_path: GPT_SoVITS/pretrained_models/s1v3.ckpt
  version: v4
  vits_weights_path: GPT_SoVITS/pretrained_models/gsv-v4-pretrained/s2Gv4.pth
v1:
  bert_base_path: GPT_SoVITS/pretrained_models/chinese-roberta-wwm-ext-large
  cnhuhbert_base_path: GPT_SoVITS/pretrained_models/chinese-hubert-base
  device: cpu
  is_half: false
  t2s_weights_path: GPT_SoVITS/pretrained_models/s1bert25hz-2kh-longer-epoch=68e-step=50232.ckpt
  version: v1
  vits_weights_path: GPT_SoVITS/pretrained_models/s2G488k.pth
v2:
  bert_base_path: /workspace/GPT_SoVITS/pretrained_models/chinese-roberta-wwm-ext-large
  cnhuhbert_base_path: /workspace/GPT_SoVITS/pretrained_models/chinese-hubert-base
  device: cpu
  is_half: false
  t2s_weights_path: /workspace/GPT_SoVITS/pretrained_models/gsv-v2final-pretrained/s1bert25hz-5kh-longer-epoch=12-step=369668.ckpt
  version: v2
  vits_weights_path: /workspace/GPT_SoVITS/pretrained_models/gsv-v2final-pretrained/s2G2333k.pth
v3:
  bert_base_path: GPT_SoVITS/pretrained_models/chinese-roberta-wwm-ext-large
  cnhuhbert_base_path: GPT_SoVITS/pretrained_models/chinese-hubert-base
  device: cpu
  is_half: false
  t2s_weights_path: GPT_SoVITS/pretrained_models/s1v3.ckpt
  version: v3
  vits_weights_path: GPT_SoVITS/pretrained_models/s2Gv3.pth

# STEP 5
建立 /opt/GPT-SoVITS/reference 資料夾，並準備 5 ~ 10 秒的範例音檔放在 reference 資料夾底下，假設檔名為 sample.wav，這個在後續 API 會使用到。

修改docker-compose.yaml:
version: '3.8'

services:
  gpt-sovits:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: gpt-sovits-container
    environment:
      - is_half=False
      - is_share=False
      - PYTHONUNBUFFERED=1
    volumes:
      - ./output:/workspace/output
      - ./logs:/workspace/logs
      - ./SoVITS_weights:/workspace/SoVITS_weights
      - ./GPT_weights:/workspace/GPT_weights
      - ./reference:/workspace/reference
    command:
      - /workspace/venv/bin/python
      - api_v2.py
      - -a
      - 0.0.0.0
      - -p
      - "9880"
      - -c
      - GPT_SoVITS/configs/tts_infer.yaml
    working_dir: /workspace
    ports:
      - "9880:9880"
      - "9871:9871"
      - "9872:9872"
      - "9873:9873"
      - "9874:9874"
    shm_size: 16G
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: "all"
              capabilities: [gpu]
    stdin_open: true
    tty: true
    restart: unless-stopped

#构建镜像：
docker build --no-cache --build-arg IMAGE_TYPE=full -t gpt-sovits:latest .

#运行容器：
docker-compose up --build -d

# 查看容器中 API Log
docker logs --follow gptsovits

# 如果正常執行沒問題，您應該可以在終端機看到以下內容（version : v4，官方设置默认不是调用最新v4模型）
---------------------------------------------TTS Config---------------------------------------------
device              : cuda
is_half             : True
version             : v4
t2s_weights_path    : GPT_SoVITS/pretrained_models/s1v3.ckpt
vits_weights_path   : GPT_SoVITS/pretrained_models/gsv-v4-pretrained/s2Gv4.pth
bert_base_path      : GPT_SoVITS/pretrained_models/chinese-roberta-wwm-ext-large
cnhuhbert_base_path : GPT_SoVITS/pretrained_models/chinese-hubert-base
----------------------------------------------------------------------------------------------------
... (省略)...
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:9880 (Press CTRL+C to quit)
其他可能會使用到的 docker 指令。

# 停止容器
docker stop gptsovits

# 刪除容器
docker rm gptsovits

# 刪除映像檔
docker rmi gptsovits

# 清除建立映像檔時的快取
docker builder prune
測試 API
在瀏覽器輸入 http://localhost:9880/docs，可以查看 FastAPI 開啟的所有 API 資訊，我們可以使用 [GET] /tts 進行測試。重要參數如下:

text
要合成的語音文字。
text_lang
合成語音文字語系代碼，中文填寫 zh，英文填寫 en。
ref_audio_path
參考範例音檔，就是放在 reference 資料夾中的音檔路徑 (/workspace/reference/sample.wav)。
prompt_lang
範例音檔案語系，中文填寫 zh，英文填寫 en。
prompt_text
範例音檔案對應的文字。
除了以上參數，其他都用預設值即可，點選下方的 Execute 按鈕，複製 Request URL 貼到瀏覽器網址列，如果系統架設沒問題的話，就能看到合成的音檔。
